### PR TITLE
Dialyzer fix

### PR DESCRIPTION
**Pull request by terry-xiaoyu**

### CODE REVIEW

**Comment by zmstone** [comment]
there is typo in the commit log for 801b995

File: apps/emqx_rule_engine/src/emqx_rule_actions.erl [hunk 1]
```diff
@@ -204,7 +203,7 @@ increase_and_publish(ActId, Msg) ->
     emqx_metrics:inc_msg(Msg).
 
 -spec on_action_create_do_nothing(action_instance_id(), Params :: map())
-      -> NewParams :: map().
+      -> {bindings(), NewParams :: map()}.
```
Comment (zmstone) [inline apps/emqx_rule_engine/src/emqx_rule_actions.erl:206]: NIT: I have seen a lot of places having `->`  leading the line, elvis will probably not rock.
what kind of convention are we following ?

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl [hunk 2]
```diff
@@ -72,7 +72,7 @@
 %% Load all providers .
 -spec(load_providers() -> ok).
 load_providers() ->
-    [load_provider(App) || App <- ignore_lib_apps(application:loaded_applications())],
+    _ = [load_provider(App) || App <- ignore_lib_apps(application:loaded_applications())],
```
Comment (zmstone) [inline apps/emqx_rule_engine/src/emqx_rule_engine.erl:75]: `lists:foreach` for the win

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl [hunk 3]
```diff
@@ -86,7 +86,7 @@ load_provider(App) when is_atom(App) ->
 %% Load all providers .
 -spec(unload_providers() -> ok).
 unload_providers() ->
-    [unload_provider(App) || App <- ignore_lib_apps(application:loaded_applications())],
+    _ = [unload_provider(App) || App <- ignore_lib_apps(application:loaded_applications())],
```
Comment (zmstone) [inline apps/emqx_rule_engine/src/emqx_rule_engine.erl:89]: `lists:foreach` ftw

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl [hunk 4]
```diff
@@ -139,24 +139,23 @@ new_resource_type({App, Mod, #{name := Name,
                    description = maps:get(description, Params, ?descr)}.
 
 find_attrs(App, Def) ->
-    [{App, Mod, Attr} || {ok, Modules} <- [application:get_key(App, modules)],
+    _ = [{App, Mod, Attr} || {ok, Modules} <- [application:get_key(App, modules)],
```
Comment (zmstone) [inline apps/emqx_rule_engine/src/emqx_rule_engine.erl:142]: isn't this a legit return value ?

Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine.erl:142]: I'lll fix it

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl (hunk 2 above)
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine.erl:75]: OK

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl (hunk 3 above)
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine.erl:89]: OK

File: apps/emqx_rule_engine/src/emqx_rule_actions.erl (hunk 1 above)
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_actions.erl:206]: No conventions for that, I'll follow the guide of elvis.

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl [hunk 5]
```diff
@@ -349,7 +351,7 @@ refresh_resource_status() ->
 %% Internal Functions
 %%------------------------------------------------------------------------------
 prepare_actions(Actions, NeedInit) ->
-    [prepare_action(Action, NeedInit) || Action <- Actions].
+    _ = [prepare_action(Action, NeedInit) || Action <- Actions].
```
[the remark this answers is not in the document]
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine.erl:354]: OK

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl [hunk 6]
```diff
@@ -414,7 +416,7 @@ ignore_lib_apps(Apps) ->
                common_test, observer, webtool, xmerl, tools,
                test_server, compiler, debugger, eunit, et,
                wx],
-    [AppName || {AppName, _, _} <- Apps, not lists:member(AppName, LibApps)].
+    _ = [AppName || {AppName, _, _} <- Apps, not lists:member(AppName, LibApps)].
```
[the remark this answers is not in the document]
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine.erl:419]: OK

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl [hunk 7]
```diff
@@ -299,6 +298,7 @@ ensure_resource_deleted(ResId) ->
 
 -spec(refresh_resources() -> ok).
 refresh_resources() ->
+    _ =
```
[the remark this answers is not in the document]
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine.erl:301]: Yes you are right, I just have forgotten the changing these to lists:foreach-es.

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl [hunk 8]
```diff
@@ -311,14 +311,16 @@ refresh_resources() ->
...
 refresh_resource(Type) when is_atom(Type) ->
+    _ =
     [refresh_resource(Resource)
      || Resource <- emqx_rule_registry:get_resources_by_type(Type)];
 refresh_resource(#resource{id = ResId, config = Config, type = Type}) ->
     {ok, #resource_type{on_create = {M, F}}} = emqx_rule_registry:find_resource_type(Type),
     cluster_call(init_resource, [M, F, ResId, Config]).
 
--spec(refresh_rules() -> ok).
+-spec(refresh_rules() -> no_return()).
```
[the remark this answers is not in the document]
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine.erl:321]: I'll fix it.

File: apps/emqx_rule_engine/src/emqx_rule_engine.erl [hunk 9]
```diff
@@ -228,7 +227,7 @@ create_resource(#{type := Type, config := Config} = Params) ->
             {error, {resource_type_not_found, Type}}
     end.
 
--spec(start_resource(resource_id()) -> ok | {error, Reason :: term()}).
+-spec(start_resource(resource_id()) -> ok | {error, Reason :: term()} | no_return()).
```
[the remark this answers is not in the document]
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine.erl]: Fixed by make sure the function always return (does not crash).

**Pull request merged**